feat: add phantom completion

timfjord opened the pull request:
<img width="959" alt="image" src="https://user-images.githubusercontent.com/471335/181252209-7c70779d-a90a-4318-b99d-7ab234adc190.png">

- [x] hide phantoms `on_selection_modified_async`
- [x] handle multi-line display text
- [x] make the completion method configurable
- [x] clean up the phantoms dict when a view is closed
- [x] try `line-height: 0` for the empty block phantom we use to keep the cursor from the left

TODOs:
- [ ] update the currently displayed phantom(instead of making a new request) when typing characters that match the beginning of the currently displayed phantom (this should work even when `auto_ask_completions` is set to `false`)
- [ ] split the display text into regions and display with multiple phantoms
- [ ] fix corner cases when the cursor jumps when the completion at the end of the line
- [ ] do not ask for completion when saving a view with `lsp_format_on_save`
- [ ] do not ask for completion after `undo`/`redo`
- [ ] [optional] override the background colour for the block phantom and make it the same as the background colour of the editor (try [this workaround](https://stackoverflow.com/a/52440845/1078179) to make it full width), this way it will be different from the `line_highlight` colour

jfcherng commented:
@TheSecEng has something similar or superior iirc

TheSecEng commented:
@timfjord this is what I was working on in a local branch.

![image](https://user-images.githubusercontent.com/32599364/181127521-ab1a5a43-bbb4-4320-85d8-07933e6f592a.png)


Is this something that is worth pursuing?

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -83,6 +84,7 @@ def hide(self) -> None:
         # prevent from hiding other plugin's popup
         if self.is_visible:
             _PopupCompletion.hide(self.view)
+            _PhantonCompletion.hide(self.view)
```
typo: Phanto`m`

timfjord replied:
🤦🏻

timfjord commented:
> @timfjord this is what I was working on in a local branch.
> 
> ![image](https://user-images.githubusercontent.com/32599364/181127521-ab1a5a43-bbb4-4320-85d8-07933e6f592a.png)
> 
> Is this something that is worth pursuing?

Absolutely @TheSecEng 

I just started on the implementation and there are a lot of things to do to make the integration really nice.  
So if you have something please share it, so we could then combine some ideas

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -118,7 +121,22 @@ def _tidy_completion_index(self, index: int) -> int:
+    def show(self) -> None:
+        pass
+
+    @classmethod
+    @abstractmethod
```
```suggestion
    @abstractclassmethod
```

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -118,7 +121,22 @@ def _tidy_completion_index(self, index: int) -> int:
 
 
-class _PopupCompletion:
+class BaseCompletion(metaclass=ABCMeta):
```
```suggestion
class BaseCompletion(ABC):
```

And `from abc import ABC`

jfcherng replied:
Sorry, forget we are at py33.

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -83,6 +84,7 @@ def hide(self) -> None:
         # prevent from hiding other plugin's popup
         if self.is_visible:
             _PopupCompletion.hide(self.view)
+            _PhantonCompletion.hide(self.view)
```
```suggestion
            _PhantomCompletion.hide(self.view)
```

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -105,6 +107,7 @@ def show(
             return
 
         _PopupCompletion(self.view).show()
+        _PhantonCompletion(self.view).show()
```
```suggestion
        _PhantomCompletion(self.view).show()
```

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -259,3 +273,78 @@ def show(self) -> None:
+_phantom_sets_per_view = {}
+
+
+class _PhantonCompletion(BaseCompletion):
```
```suggestion
class _PhantomCompletion(BaseCompletion):
```

TheSecEng commented:
> I've updated the integration a bit, so it now handles multi-line strings and properly ident everything with `&nbsp;` symbols.
> 
> The cursor still jumps in some cases and for now, I am not sure how to fix it(my workaround with an empty block doesn't work in all cases) but in general, it is now a way more reliable

I wonder if we just reset cursor position after drawing the phantom would work.

TheSecEng commented:
Also, I am assuming we plan to put this behind a setting? So the user can select popup vs phantom?

timfjord commented:
> > I've updated the integration a bit, so it now handles multi-line strings and properly ident everything with `&nbsp;` symbols.
> > The cursor still jumps in some cases and for now, I am not sure how to fix it(my workaround with an empty block doesn't work in all cases) but in general, it is now a way more reliable
> 
> I wonder if we just reset cursor position after drawing the phantom would work.

I haven't tried this, but the problem with phantom cursor jumping is that the position doesn't really change.  
I think it is basically how Sublime handles phantom rendering.  
I am going to play with it a bit more, but most likely we need to ask someone from the Sublime team (I might try in the discord chat if I won't be able to find a solution).

So as far as I understand, you are also experiencing these problems with the cursor in your implementation?

TheSecEng commented:
> > I wonder if we just reset cursor position after drawing the phantom would work.
> 
> I haven't tried this, but the problem with phantom cursor jumping is that the position doesn't really change. I think it is basically how Sublime handles phantom rendering. I am going to play with it a bit more, but most likely we need to ask someone from the Sublime team (I might try in the discord chat if I won't be able to find a solution).
> 
> So as far as I understand, you are also experiencing these problems with the cursor in your implementation, right?

Yup, and there is no rhyme or reason to when it happens.

I experienced this and an old plugin as well.

TheSecEng commented:
For settings, we should really access `session.config.settings` over `view.settings()` as the session holds the most true settings for each LSP session

TheSecEng commented:
@timfjord Perhaps this is what we experience https://github.com/sublimehq/sublime_text/issues/2871

timfjord commented:
I think that is exactly it, let me add my comment there.  
A label was added on May'31, so at least it was acknowledged by the Sublime team

TheSecEng commented:
There is also https://github.com/sublimehq/sublime_text/issues/4575 which might be relevant

timfjord commented:
<img width="637" alt="image" src="https://user-images.githubusercontent.com/471335/181456672-2cde7498-44fd-46a6-8112-e010913bc1cf.png">

It was tagged on the same day as the other one, so maybe there is chance it will be implemented

jfcherng commented on plugin/listeners.py:
```diff
@@ -8,10 +8,18 @@
 
 class ViewEventListener(sublime_plugin.ViewEventListener):
+    @property
+    def _is_modifed(self) -> bool:
```
typo: \_is_modif`i`ed

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -274,3 +295,94 @@ def show(self) -> None:
+        completion = self.completion_manager.current_completion
+        assert completion
+
+        head, *tail = completion["displayText"].split("\n")
```
`str.partition` is perfect for this use. It should be faster and more memory efficient (not important here though 😄 )

timfjord replied:
Unfortunately `str.partition` won't work, because I need tot split text into an array, so I can render each line separately

jfcherng replied:
> Unfortunately `str.partition` won't work here, because I need to split the text into a list, so I can render each line separately

So you handle the second line in a different way with the third line? I don't expect that. I think only the first and the rest are different.

timfjord replied:
Basically I need a list of lines. 
The first line is rendered as an inline phantom and all other lines are rendered as a block phantom but each line needs to be wrapped in `<div class="copilot-completion-line>`, that's why I need the `split`

jfcherng replied:
![image](https://user-images.githubusercontent.com/6594915/182427284-535c56c8-06b6-4866-8820-47db782b63c5.png)

From my experiment, text in a phantom also auto follows `line_padding_top` and `line_padding_bottom`. There is no need for CSS.

timfjord replied:
It follows only in the inline phantom(first line), and I don't add padding there.  
But, unfortunately, it doesn't follow in the block phantom that is displayed underneath

jfcherng replied:
> But, unfortunately, it doesn't follow in the block phantom that is displayed underneath

ah, I see that now...

timfjord commented:
@jfcherng thanks to your latest debounce improvements the cursor doesn't jump anymore 👍🏻   
I think the main issue with the cursor was that we sent too many requests(because of the old debouncer implementation) and the phantoms were shown and hidden multiple times that was causing the cursor to jump.

There might be still problems with cursor jumping but at least it is way more stable now

jfcherng commented:
Personally, I have `"line_padding_bottom": 2,` in my preferences.

There is a similar issue to https://github.com/facelessuser/ColorHelper/pull/122.

https://user-images.githubusercontent.com/6594915/182443500-8319948d-228b-4715-8f77-f0fb2a7d1a3c.mp4

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -274,3 +295,94 @@ def show(self) -> None:
+
+        self._phantom_set.update(
+            [
+                self._build_phantom(head, completion["point"] + 1, completion["point"]),
```
Is there a magic? Why the `end` is smaller than `begin`?

timfjord replied:
Yeah, it is a workaround to keep the cursor from the right(unfortunately it doesn't always work, but it is better than nothing)  
There is a note about the empty block, but I will add a note about this trick too

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -274,3 +295,94 @@ def show(self) -> None:
+    <body id="copilot-completion">
+        <style>
+            body {{
+                color: #808080;
```
Should we use `comment` color for this? 

E.g.,
```py
>>> view.style_for_scope('comment')
{'bold': False, 'foreground': '#616161', 'italic': False, 'selection_foreground': '#757575', ...}
```

timfjord replied:
I don't know, I took this colour from [the neovim plugin](https://github.com/github/copilot.vim/blob/release/plugin/copilot.vim#L14-L20) and it works ok for both light and dark modes.

What I would do I would add a note on how to customize it, but, IMO, it sounds like a reasonable default

jfcherng replied:
If users don't complain in the neovim plugin, I guess we don't have to do that until many of them complain.

timfjord commented:
> Personally, I have `"line_padding_bottom": 2,` in my preferences.
> 
> There is a similar issue to [facelessuser/ColorHelper#122](https://github.com/facelessuser/ColorHelper/pull/122).
> 
>  dasf.mp4

An empty block is added to stabilize the cursor and  it causes the current line to be a big bigger.  
I haven't tried `line-height: 0`, but  I will do

jfcherng commented on plugin/ui/completion.py:
```diff
@@ -274,3 +298,95 @@ def show(self) -> None:
+        assert self._phantom_set
+        self._phantom_set.update(
+            [
+                self._build_phantom([first_line], completion["point"] + 1, completion["point"]),
```
```suggestion
                self._build_phantom(first_line, completion["point"] + 1, completion["point"]),
```

Just notice that I forgot to remove the brackets.

timfjord commented:
@jfcherng @TheSecEng all the edge cases for the phantom completion are taking longer than planned so I would like to propose the following.

I've addressed the most critical TODOs so to avoid maintaining a long-running branch I would send this one to review.
As for the remaining TODOs, I would separate issues and tackle them separately.
What do you think?

I've also done minor refactoring as part of this PR. 
Basically, I've made the Completion classes more independent and removed `ViewCompletionManager` from its dependencies.

timfjord commented on plugin/ui/completion.py:
```diff
@@ -84,24 +132,46 @@ def show_next_completion(self) -> None:
+
+        self.hide()
+
+    def handle_text_change(self) -> None:
```
I know that this is is exacrtly the same as the one above, but I am planning to extend it in a follow-up and handle text changes when the phantom completion is visible

timfjord commented:
Ok, I think it is ready to go

timfjord commented:
Hey @TheSecEng 
Have you got a chance to look at this PR?
